su1/worker-pool.go

```diff
@@ -0,0 +1,63 @@
+package su1
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+type Task func() error
+
+func WorkerPool(ctx context.Context, maxWorkers int, tasks <-chan Task) (<-chan struct{}, <-chan error) {
+	var wg sync.WaitGroup
+
+	done := make(chan struct{})
+	errors := make(chan error, maxWorkers)
+
+	worker := func(id int) {
+		fmt.Printf("Starting instance %d\n", id)
+
+		defer (func() {
+			fmt.Printf("Stopping instance %d\n", id)
+
+			wg.Done()
+		})()
+
+		for {
+			select {
+			case <-ctx.Done():
+				fmt.Printf("Worker %d stopped due to context cancellation\n", id)
+				return
+			case task, ok := <-tasks:
+				if !ok {
+					fmt.Printf("Worker %d stopped due to closed channel\n", id)
+					return
+				}
+
+				err := task()
+
+				if err != nil {
+					select {
+					case errors <- err:
```

Comment:
Channel is a synchronization tool. If we define that we return two channels and the client can read from both - it is an API contract. Here, according to the `default` instruction, the worker pool will keep working even if the client ignores them. 

The client should not ignore errors that are defined in the contract. It is not the worker pool's responsibility to solve the client's invalid API usage. 

We either send errors to a channel and lock or just gather all errors into a slice and return them at the end.